src/main/scala/org/renci/cam/QueryService.scala

```diff
@@ -181,11 +189,83 @@ object QueryService extends LazyLogging {
+      kgEdges <- Task.effect(mutable.ListBuffer[TranslatorEdge]())
+      prefixes <- Task.effect(PREFIXES.map(entry => s"PREFIX ${entry._1}: <${entry._2}>").mkString("\n"))
+      results <- ZIO.foreach(resultSet.asScala.toList) { querySolution =>
+        for {
+          nodeMap <- Task.effect(queryGraph.nodes.map(n => (n.id, querySolution.get(s"${n.id}_type").toString)).toMap)
+          nodeBindings <- ZIO.foreach(queryGraph.nodes) { n =>
+            for {
+              abbreviatedNodeType <- Task.effect(applyPrefix(nodeMap.get(n.id).get))
+              nodeDetails <- getKnowledgeGraphNodeDetails(prefixes, String.format("<%s>", nodeMap.get(n.id).get))
+              nodeDetailsHead <- Task.effect(nodeDetails.head)
+              _ = kgNodes += TranslatorNode(abbreviatedNodeType, Some(nodeDetailsHead._1), Some(nodeDetailsHead._2), None)
+              nodeBinding <- Task.effect(TranslatorNodeBinding(n.id, abbreviatedNodeType))
+            } yield nodeBinding
+          }
+
+          edgeBindings <- ZIO.foreach(queryGraph.edges) { e =>
+            for {
+              predicateRDFNode <- Task.effect(querySolution.get(e.id).toString)
+              sourceRDFNode <- Task.effect(querySolution.get(e.source_id).toString)
+              targetRDFNode <- Task.effect(querySolution.get(e.target_id).toString)
+              knowledgeGraphId = {
+                val newTranslatorEdge =
+                  NewTranslatorEdge(e.`type`, e.source_id, e.target_id).asJson.deepDropNullValues.noSpaces.getBytes(StandardCharsets.UTF_8)
+                val messageDigest = MessageDigest.getInstance("SHA-256")
+                String.format("%064x", new BigInteger(1, messageDigest.digest(newTranslatorEdge)))
+              }
+              _ = kgEdges += TranslatorEdge(knowledgeGraphId,
+                                            Some(e.`type`),
+                                            applyPrefix(nodeMap.get(e.source_id).get),
+                                            applyPrefix(nodeMap.get(e.target_id).get))
+              prov <- getProvenance(prefixes, sourceRDFNode, predicateRDFNode, targetRDFNode)
+            } yield TranslatorEdgeBinding(e.id, knowledgeGraphId, Some(prov.toString))
+          }
+
+        } yield TranslatorResult(nodeBindings, edgeBindings)
+      }
+    } yield TranslatorMessage(Some(queryGraph), Some(TranslatorKnowledgeGraph(kgNodes.toList, kgEdges.toList)), Some(results))
+
+  def getProvenance(prefixes: String, source: String, predicate: String, target: String): RIO[Config[AppConfig], String] =
+    for {
+      query <- Task.effect(QueryFactory.create(s"""$prefixes
+SELECT ?g ?other WHERE { GRAPH ?g { <$source> <$predicate> <$target> } OPTIONAL { ?g prov:wasDerivedFrom ?other . }
+}"""))
+      bindings <- SPARQLQueryExecutor.runSelectQuery(query)
+      nextSolution = bindings.nextSolution()
+      prov <- Task.effect(nextSolution.get("g").toString).orElse(Task.effect(nextSolution.get("other").toString))
```

Comment:
I think this is backwards. If `?other` is bound, then that means the triple was found in an inferred graph, which was derived from the graph in `?other`. So take `?other` first, if it is not bound then take `?g` (meaning the triple is in the asserted graph).

Reply:
Easy enough to fix.